go/teams/hidden/hidden.go

```diff
@@ -0,0 +1,565 @@
+// endpoint tells us otherwise.
+func (l *LoaderPackage) SetIsFresh(b bool) {
+	l.isFresh = b
+}
+
+// IsFresh returns false only if someone called SetIsFresh(false).
+func (l *LoaderPackage) IsFresh() bool {
+	return l.isFresh
+}
+
+func (l *LoaderPackage) checkPrev(mctx libkb.MetaContext, first sig3.Generic) (err error) {
+	q := first.Seqno()
+	prev := first.Prev()
+	if q == keybase1.Seqno(1) && prev != nil {
+		return NewLoaderError("bad link that had seqno=1 and non=nil prev")
+	}
+	if prev == nil {
+		return nil
+	}
+	if l.data == nil {
+		return NewLoaderError("didn't get prior data and update was for a chain middle")
+	}
+	link, ok := l.data.Outer[q-1]
+	if !ok {
+		return NewLoaderError("previous link wasn't found")
+	}
+	if !link.Eq(prev.Export()) {
+		return NewLoaderError("prev mismatch at %d", q)
+	}
+
+	// We check prevs again when we commit this change to the hidden team chain manager. It's not
+	// strictly required, but it seems a good safeguard against future programming bugs. So
+	// store it away here.
+	l.expectedPrev = &keybase1.LinkTriple{
+		Seqno:   q - 1,
+		LinkID:  link,
+		SeqType: keybase1.SeqType_TEAM_PRIVATE_HIDDEN,
+	}
+
+	return nil
+}
+
+func (l *LoaderPackage) checkExpectedHighSeqno(mctx libkb.MetaContext, links []sig3.Generic) (err error) {
+	last := l.LastSeqno()
+	max := l.MaxRatchet()
+	if max <= last {
+		return nil
+	}
+	if len(links) > 0 && links[len(links)-1].Seqno() >= max {
+		return nil
+	}
+	return NewLoaderError("Server promised a hidden chain up to %d, but never recevied; is it withholding?", max)
+}
+
+func (l *LoaderPackage) checkRatchet(mctx libkb.MetaContext, update *keybase1.HiddenTeamChain, ratchet keybase1.LinkTripleAndTime) (err error) {
+	q := ratchet.Triple.Seqno
+	link, ok := update.Outer[q]
+	if ok && !link.Eq(ratchet.Triple.LinkID) {
+		return NewLoaderError("update data failed to match ratchet %+v v %s", ratchet, link)
+	}
+	return nil
+}
+
+func (l *LoaderPackage) checkRatchets(mctx libkb.MetaContext, update *keybase1.HiddenTeamChain) (err error) {
+	if l.data == nil {
+		return nil
+	}
+	for _, r := range l.data.Ratchet.Flat() {
+		err = l.checkRatchet(mctx, update, r)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Update combines the preloaded data with any downloaded updates from the server, and stores
+// the result local to this object.
+func (l *LoaderPackage) Update(mctx libkb.MetaContext, update []sig3.ExportJSON) (err error) {
+	defer mctx.Trace(fmt.Sprintf("LoaderPackage#Update(%s)", l.id), func() error { return err })()
+
+	var data *keybase1.HiddenTeamChain
+	data, err = l.updatePrecheck(mctx, update)
+	if err != nil {
+		return err
+	}
+	err = l.storeData(mctx, data)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (l *LoaderPackage) updatePrecheck(mctx libkb.MetaContext, update []sig3.ExportJSON) (ret *keybase1.HiddenTeamChain, err error) {
+	var links []sig3.Generic
+	links, err = importChain(mctx, update)
```

Comment:
It would be nice if importChain and checkLinkSequence were both in sig3 so it's harder to get a malformed sig3 chain.

Reply:
the point of `importChain` is to import to `sig3`s?